### ISSUE TITLE
Max String length 16 when setting OpenAPI schema constraint maxLength to a higher value

### ISSUE THREAD

**Issue by chrstphlbr** [EMResearch/EvoMaster#803]
Hi,

I am using EvoMaster 1.6.1, which I have compiled from source code (checking out the git tag v1.6.1). I tried setting an OpenAPI schema string constraint to maxLength 20. However, I am getting an exception with the message "java.lang.IllegalArgumentException: Min 20 is bigger than max 16" (see adapted stacktrace below).

I assume that when OpenAPI constraints are used, the EMConfig field `maxLengthForStringsAtSamplingTime` or the line 177 in the method `org.evomaster.core.search.gene.string.StringGene.randomize` is not properly set and the maxLength from the constraints is overridden/not considered.

Let me know if you need any more information or if this bug has been resolved already.

Christoph

Adapted stacktrace:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException: invokedynamic: method=public void org.evomaster.core.problem.rest.service.AbstractRestSampler.initialize(), target=CustomResourceSampler@2ec3633f
        at com.netflix.governator.lifecycle.LifecycleMethods.methodInvoke(LifecycleMethods.java:316)
        at com.netflix.governator.lifecycle.LifecycleMethods.methodInvoke(LifecycleMethods.java:298)
        at com.netflix.governator.lifecycle.LifecycleManager.startInstance(LifecycleManager.java:266)
        at com.netflix.governator.lifecycle.LifecycleManager.add(LifecycleManager.java:175)
        at com.netflix.governator.guice.InternalLifecycleModule.onProvision(InternalLifecycleModule.java:81)
        at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:117)
        at com.google.inject.internal.ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:93)
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:296)
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
        at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:169)
        at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
        at com.google.inject.internal.FactoryProxy.get(FactoryProxy.java:60)
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
        at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:169)
        at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
        at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:213)
        at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:186)
        at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:113)
        at com.google.inject.internal.InjectorImpl.createChildInjector(InjectorImpl.java:239)
        at com.netflix.governator.guice.LifecycleInjector.createChildInjector(LifecycleInjector.java:331)
        at com.netflix.governator.guice.LifecycleInjector.createInjector(LifecycleInjector.java:411)
        at com.netflix.governator.guice.LifecycleInjector.createInjector(LifecycleInjector.java:352)
        at CustomEMSetup.injector(CustomEMSetup.kt:68)
        at CustomMainKt.main(Main.kt:33)
        at CustomMainKt..main(Main.kt)
Caused by: java.lang.IllegalArgumentException: Min 20 is bigger than max 16
        at org.evomaster.core.search.service.Randomness.nextInt(Randomness.kt:157)
        at org.evomaster.core.search.service.Randomness.nextWordString(Randomness.kt:207)
        at org.evomaster.core.search.gene.string.StringGene.randomize(StringGene.kt:186)
        at org.evomaster.core.search.gene.ObjectGene.randomize(ObjectGene.kt:109)
        at org.evomaster.core.search.gene.collection.ArrayGene.createRandomElement(ArrayGene.kt:314)
        at org.evomaster.core.search.gene.collection.ArrayGene.randomize(ArrayGene.kt:186)
        at org.evomaster.core.search.gene.optional.SelectableWrapperGene.randomize(SelectableWrapperGene.kt:59)
        at org.evomaster.core.search.gene.ObjectGene.randomize(ObjectGene.kt:109)
        at org.evomaster.core.search.gene.Gene.doInitialize(Gene.kt:171)
        at org.evomaster.core.search.Action.doInitialize(Action.kt:73)
        at org.evomaster.core.problem.rest.resource.RestResourceNode.sampleOneAction(RestResourceNode.kt:392)
        at org.evomaster.core.problem.rest.service.ResourceManageService.createAdHocIndividuals(ResourceManageService.kt:103)
        at CustomResourceSampler.customizeAdHocInitialIndividuals(CustomResourceSampler.kt:30)
        at org.evomaster.core.problem.rest.service.AbstractRestSampler.initAdHocInitialIndividuals(AbstractRestSampler.kt:170)
        at org.evomaster.core.problem.rest.service.AbstractRestSampler.initForBlackBox(AbstractRestSampler.kt:277)
        at org.evomaster.core.problem.rest.service.AbstractRestSampler.initialize(AbstractRestSampler.kt:60)
        at com.netflix.governator.lifecycle.LifecycleMethods.methodInvoke(LifecycleMethods.java:313)
        ... 25 more
```

**Comment by arcuri82** [comment]
Hi @chrstphlbr 
 thx. It indeed looks like a bug when the `min` length declared in the schema is greater than 16.

btw, why did you close the issue?

**Comment by chrstphlbr** [comment]
I closed it, because I was unsure whether I opened it prematurely before understanding whether I made an error implementing a custom sampler. I will have a closer look today and update the issue in case I find something.

**Comment by chrstphlbr** [comment]
I looked at it in more detail. The culprit seems to be in `org.evomaster.core.search.gene.string.StringGene.randomize` when `adjustedMin` is larger than `adjustedMax` and `maxForRandomization` defaults to 16.

This happens in my custom RestSampler where I first randomly sample and initialize a `RestCallAction` and afterwards "attach" it to a `RestIndividual`. However, this pattern (first create an unattached RestCallAction -> initialize -> create Individual) also occurs (from where I got the inspiration for my custom RestSampler) in `RestSampler` and `ResourceSampler`. I assume, therefore, that EvoMaster's built-in REST samplers are also affected by this bug.

As `maxForRandomization` defaults to 16 in this case, as `getSearchGlobalState()` returns null as the RestCallAction does not have an `Individual` as a parent yet, I can not even circumvent the bug by setting the `EMConfig.maxLengthForStringsAtSamplingTime`.

I hope this additional information helps.
